### PR TITLE
`gpu(::DataLoader)`, take III

### PR DESCRIPTION
**Pull request by mcabbott**
Simpler variant that just calls `DataLoader(mapobs(gpu, data), ...)`.

What this misses compared to more complex CuIterator-like things is that it does not call `finalize` afterwards. But perhaps that doesn't matter since each call of the model will allocate so much more, and also not `finalize` explicitly.

Closes #2240, closes #2186

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

### CODE REVIEW

File: docs/src/gpu.md [hunk 1]
```diff
@@ -122,61 +122,48 @@ julia> x |> cpu
...
-Some of the common workflows involving the use of GPUs are presented below.
-
-### Transferring Training Data
+## Transferring Training Data
 
-In order to train the model using the GPU both model and the training data have to be transferred to GPU memory. This process can be done with the `gpu` function in two different ways:
+In order to train the model using the GPU both model and the training data have to be transferred to GPU memory. Moving the data can be done in two different ways:
 
-1. Iterating over the batches in a [DataLoader](@ref) object transferring each one of the training batches at a time to the GPU. 
+1. Iterating over the batches in a [`DataLoader`](@ref) object transferring each one of the training batches at a time to the GPU. This is recommended for large datasets. Done hand, it might look like this:
```
Comment (CarloLucibello) [inline docs/src/gpu.md:129]: "done hand"?

File: docs/src/gpu.md [hunk 2]
```diff
@@ -122,61 +122,48 @@ julia> x |> cpu
...
-   ```
-   Note that both `gpu` and `cpu` are smart enough to recurse through tuples and namedtuples. Another possibility is to use [`MLUtils.mapsobs`](https://juliaml.github.io/MLUtils.jl/dev/api/#MLUtils.mapobs) to push the data movement invocation into the background thread:
-   ```julia
-   using MLUtils: mapobs
-   # ...
-   gpu_train_loader = Flux.DataLoader(mapobs(gpu, (xtrain, ytrain)), batchsize = 16)
-   ```
-
-3. Wrapping the `DataLoader` in [`CUDA.CuIterator`](https://cuda.juliagpu.org/stable/usage/memory/#Batching-iterator) to efficiently move data to GPU on demand:
+   Rather than write this every time, there is a method of `gpu(::DataLoader)` which does it for you:
```
Comment (CarloLucibello) [inline docs/src/gpu.md:142]:
```suggestion
   Rather than write this every time, the method `gpu(::DataLoader)` does it for you:
```

File: docs/src/gpu.md [hunk 3]
```diff
@@ -122,61 +122,48 @@ julia> x |> cpu
...
+   # ... model definition, optimiser setup
+   for epoch in 1:epochs
+       for (x, y) in gpu_train_loader
+           grads = gradient(m -> loss(m, x, y), model)
+           Flux.update!(opt_state, model, grads[1])
        end
    end
    ```
+   This is equivalent to `DataLoader(MLUtils.mapobs(gpu, (X, Y)); keywords...)`.
+   Something similar can also be done with [`CUDA.CuIterator`](https://cuda.juliagpu.org/stable/usage/memory/#Batching-iterator), `gpu_train_loader = CUDA.CuIterator(train_loader)`. However, this only works with a limited number of data types: `first(train_loader)` should be a tuple (or `NamedTuple`) of arrays.
```
Comment (CarloLucibello) [inline docs/src/gpu.md:154]: Here we could hint at using `mapobs` to transform the dataset into something CUIterator compatible. Could be a short example like
```julia
train_loader = mapobs(preprocess_transform, train_loader)
gpu_train_loader = CUDA.CuIterator(train_loader)
```
 Also, mention when CuIterator should be preferred over gpu?

File: docs/src/gpu.md [hunk 4]
```diff
@@ -122,61 +122,48 @@ julia> x |> cpu
...
-   Note that this works with a limited number of data types. If `iterate(train_loader)` returns anything other than arrays, approach 1 or 2 is preferred.
+2. Transferring all training data to the GPU at once before creating the `DataLoader`. This is usually performed for smaller datasets which are sure to fit in the available GPU memory.
+   ```julia
+   gpu_train_loader = Flux.DataLoader((X, Y) |> gpu, batchsize = 32)
+   # ...
+   for epoch in 1:epochs
+       for (x, y) in gpu_train_loader
+           # ...
+   ```
+   Here `(X, Y) |> gpu` applies [`gpu`](@ref) to both arrays -- it recurses into not just tuples, as here, but also whole Flux models.
```
Comment (CarloLucibello) [inline docs/src/gpu.md:164]:
```suggestion
   Here `(X, Y) |> gpu` applies [`gpu`](@ref) to both arrays -- it recurses into not just tuples, as here, but also nested data structures.
```

File: docs/src/gpu.md (hunk 1 above)
Comment (mcabbott) [reply docs/src/gpu.md:129]:
```suggestion
1. Iterating over the batches in a [`DataLoader`](@ref) object transferring each one of the training batches at a time to the GPU. This is recommended for large datasets. Done by hand, it might look like this:
```

File: docs/src/gpu.md (hunk 4 above)
Comment (mcabbott) [reply docs/src/gpu.md:164]:
```suggestion
   Here `(X, Y) |> gpu` applies [`gpu`](@ref) to both arrays, as it recurses into structures.
```

File: docs/src/gpu.md (hunk 3 above)
Comment (mcabbott) [reply docs/src/gpu.md:154]: Is it preferred? This PR takes the line that it's not... it's tricky about what types, and finalize doesn't matter. So it's mentioned here in case people are already using it. 

If finalize does matter, then we should do https://github.com/FluxML/Flux.jl/pull/2240 instead.

File: docs/src/gpu.md (hunk 2 above)
Comment (mcabbott) [reply docs/src/gpu.md:142]:
```suggestion
   Rather than write this out every time, you can just call `gpu(::DataLoader)`:
```
